Add `Client.download_socket_value`

chrisvanrun opened the pull request:
Closes: #228

Abstract socket-value handling when downloading behind a helper function: `Client.download_socket_value`.

In contrast to the earlier example, this now also handles the `images/` relative paths correctly, takes the legacy 'generic-medical-image' into account, and has a switch in place to handle JSON values correctly.

chrisvanrun commented on gcapi/client.py:
```diff
@@ -94,6 +95,11 @@ def download(
 
+            if image.dicom_image_set is not None:
+                raise NotImplementedError(
+                    "Cannot download DICOM image sets using this method"
```
Ah, good to know. Should we incorporate that? I doubt because a.) not a lot of ppl will use it as an output (if possible at all), and b.) I would have to get access to staging to test it.

jmsmkn replied:
It's there for being incorporated into GCAPI as users will want to download their archive items.

chrisvanrun commented:
Putting this to draft to implement DICOM downloading.

chrisvanrun commented:
Ran into some model-related problems with getting the images. Fix has been merged, should use models once that is pushed to staging but I'll circumvent it for now.

chrisvanrun commented:
I've added downloading for DICOM Image sets and reworked the `Image` super kind downloading such that it is more in line with how algorithms are being presented with files on the platform.

koopmant commented on gcapi/client.py:
```diff
@@ -67,7 +77,7 @@ def download(
 
         Args:
-            filename: Base filename for downloaded files. Extension is added automatically.
+            filename: Optional, base filename for downloaded files. Extension is added automatically.
```
Add `output_directory` to this list

koopmant reviewed (commented):
Looks good as far as I can tell. Two remarks.

chrisvanrun commented on gcapi/client.py:
```diff
@@ -686,6 +780,61 @@ def _fetch_algorithm_detail(self, slug: str) -> gcapi.models.Algorithm:
+                raise ValueError(
+                    f"Unexpected super_kind {value.interface.super_kind}"
+                )
+
```
I get where you are coming from with this suggestion! However, for the legacy socket(s?) the relative_path is empty. As such, we really only know the filename/location once we are in the `images.download` for those.

For value and file super kinds we know it at this point: would it make more sense to repeat the existence check once? I.e. within each if block of `"file"` and `"value"` respectively? I've now done this:

```Python
 if super_kind == "image":
    # Image values
    self.images.download(
        url=str(value.image),
        output_directory=filename,
    )
elif super_kind == "value":
    # Direct values (e.g. '42')
    if filename.exists():
        raise FileExistsError(f"File {filename} already exists")
    with open(filename, "w") as f:
        json.dump(value.value, f, indent=2)
elif super_kind == "file":
    # Values stored as files
    if filename.exists():
        raise FileExistsError(f"File {filename} already exists")
    resp = self(
        url=str(value.file),
        follow_redirects=True,
    )
    content = (
        resp.content
        if isinstance(resp, httpx.Response)
        else json.dumps(resp).encode()
    )
    with open(filename, "wb") as f:
        f.write(content)
else:
    raise ValueError(
        f"Unexpected super_kind {value.interface.super_kind}"
    )
```
